compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/NameCollision.hs

```diff
@@ -148,21 +154,50 @@ checkDataType moduleName DefDataType{..} =
                 checkName (NEnumCon moduleName dataTypeCon vconName)
 
         DataSynonym _ ->
-            checkName (NEnumType moduleName dataTypeCon)
+            checkName (NTypeSynonym moduleName dataTypeCon)
 
 checkTemplate :: MonadGamma m => ModuleName -> Template -> S.StateT NCState m ()
 checkTemplate moduleName Template{..} = do
     forM_ tplChoices $ \TemplateChoice{..} ->
         checkName (NChoice moduleName tplTypeCon chcName)
 
+checkModuleName :: MonadGamma m => Module -> S.StateT NCState m ()
+checkModuleName m = checkName (NModule (moduleName m))
+
+checkModuleTypes :: MonadGamma m => Module -> S.StateT NCState m ()
+checkModuleTypes m = do
+    forM_ (moduleDataTypes m) $ \dataType ->
+        withContext (ContextDefDataType m dataType) $
+            checkDataType (moduleName m) dataType
+    forM_ (moduleTemplates m) $ \tpl ->
+        withContext (ContextTemplate m tpl TPWhole) $
+            checkTemplate (moduleName m) tpl
+
+checkModuleFully :: MonadGamma m => Module -> S.StateT NCState m ()
+checkModuleFully m = do
+    checkModuleName m
+    checkModuleTypes m
+
+-- | Is the first module an ascendant of the second? This check
+-- is case-insensitive because name collisions are case-insensitive.
+isAscendant :: ModuleName -> ModuleName -> Bool
```

Comment:
I’m a bit confused about this. Do we have some concept of hierarchical modules in DAML-LF? I thought this worked just like in Haskell where the hierarchy doesn’t really mean anything.

Reply:
Ah I think I got it now, we need to check for collisions between module `A.B` and type `B` in module `A`. It would be good to add a comment on this.